---
title: "scRNABatchQC Report"
output: 
  html_document:
    toc: true
    toc_depth: 2
params:
  data: plotData
  displayFigure: TRUE
---

<style>
.figure p.caption {
  font-size: 1.5em;
}
</style>

```{r setup, include=FALSE}
library(RColorBrewer)
library(knitr)
library(reshape2)
library(ggplot2)
library(data.table)
library(DT)
library(RCurl)
library(htmltools)

knitr::opts_chunk$set(echo = TRUE, autodep=TRUE, cache.comments=FALSE, message=FALSE, warning=FALSE)

addLinkTag<-function(text,link) {
  result<-paste0("<a href='",link,"' target='_blank'>",text,"</a>")
  return(result)
}

figRef <- local({
  tag <- numeric()
  created <- logical()
  used <- logical()
  function(label, caption, prefix = options("figcap.prefix"), 
           sep = options("figcap.sep"), prefix.highlight = options("figcap.prefix.highlight"), trunk.eval=TRUE) {
    if(trunk.eval){ 
      i <- which(names(tag) == label)
      if (length(i) == 0) {
        i <- length(tag) + 1
        tag <<- c(tag, i)
        names(tag)[length(tag)] <<- label
        used <<- c(used, FALSE)
        names(used)[length(used)] <<- label
        created <<- c(created, FALSE)
        names(created)[length(created)] <<- label
      }
      if (!missing(caption)) {
        created[label] <<- TRUE
        paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
               " ", caption)
      } else {
        used[label] <<- TRUE
        paste(prefix, tag[label])
      }
    }
  }
})
options(figcap.prefix = "Figure", figcap.sep = ":", figcap.prefix.highlight = "**")

tabRef <- local({
  tag <- numeric()
  created <- logical()
  used <- logical()
  function(label, caption, prefix = options("tabcap.prefix"), 
           sep = options("tabcap.sep"), prefix.highlight = options("tabcap.prefix.highlight"), trunk.eval=TRUE) {
    if(trunk.eval){
      i <- which(names(tag) == label)
      if (length(i) == 0) {
        i <- length(tag) + 1
        tag <<- c(tag, i)
        names(tag)[length(tag)] <<- label
        used <<- c(used, FALSE)
        names(used)[length(used)] <<- label
        created <<- c(created, FALSE)
        names(created)[length(created)] <<- label
      }
      if (!missing(caption)) {
        created[label] <<- TRUE
        paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
              " ", caption)
      } else {
        used[label] <<- TRUE
        paste(prefix, tag[label])
      }
    }
  }
})
options(tabcap.prefix = "Table", tabcap.sep = ":", tabcap.prefix.highlight = "**")

getHeatmapHeight<-function(hdata){
  max(10, nrow(hdata) / 10)
}

isHeatmapFigure<-function(datam){
  if (is.null(datam)){
    return(FALSE)
  }
  return (ncol(datam)>1 & nrow(datam)>1)
}

getHeatmapCaption<-function(datam, feature){
  if (isHeatmapFigure(datam)){
    return (figRef(feature))
  }else{
    return (tabRef(feature))
  }
}

plotHeatmap<-function(datam,feature=c("hvg","hvgPath","PCg","PCgPath","diffg","diffgPath")){
  keytitle<-switch(feature, "hvg"="zval","hvgPath"="-logFDR","PCg"="logFC","PCgPath"="-logFDR","diffg"="logFC","diffgPath"="-logFDR")
  maintitle<-switch(feature, "hvg"="Highly variable genes","hvgPath"="Pathways enriched in HVGs","PCg"="PC-related genes","PCPath"="Pathways enriched in PC-related genes","diffg"="Differentially expressed genes","diffgPath"="Pathways enriched in DEGs")
  if (isHeatmapFigure(datam)){
    if (feature== "hvg" | feature== "PCg" | feature=="diffg"){ 
      colv=heatmapColors
	  } 
    else {
      colv=heatmapPathwayColors
    }
    heatmap.2(datam, margins = heatmapMargins,  keysize=heatmapKeysize, col=colv,main=maintitle,key.title=keytitle,key.xlab="",key.ylab="",cexCol=1)
  } 
  else {
    warning("there is only one dataset/comparison or one gene/pathway detected and heatmap figure is not generated")
	  kable(as.matrix(datam),caption=tabRef(feature, maintitle),col.names=paste0(colnames(datam),"(",keytitle,")"))
	}
}

plotSaturation = function(plotDataSat){
  p = ggplot(plotDataSat, aes(x = readsPerCell, y = saturation, col = sampleID)) + 
    geom_line() +
    theme_bw() +
    theme(axis.text.x=element_blank(), axis.ticks.x=element_blank())
  return(p)
}

```

```{r data, include=FALSE, cache=FALSE}
plotData<-params$data
sces <- plotData$sces
hvgBiologicalSimilarity <- .getBiologicalSimilarity(sces, objectName = "hvg", valueName = "zval")
hvgPathways <- .getMultiplePathways(sces, metaObjectName = "hvgPathway")
pc1geneBiologicalSimilarity <- .getBiologicalSimilarity(sces, objectName = "pc1genes", valueName = "logFC")
pc1Pathways <- .getMultiplePathways(sces, metaObjectName = "pc1Pathway")

if(length(sces) < 9){
  scolors=brewer.pal(length(sces), "Set1")
}else{
  scolors=rainbow(length(sces))
}
names(scolors)<-names(sces)

hasComparison <- length(sces) > 1 && plotData$pwComparison

isInteractive <- length(sces) > 6

hasSaturation <- ifelse(is.null(plotData$saturation), F, T)

heatmapKeysize<-1
heatmapColors<-bluered(75)
heatmapPathwayColors<-colorpanel(75,low="white",high="red")
heatmapMargins<-c(10,15)

hasPathway<-!is.null(hvgPathways)

hasPC1Pathway<-!is.null(pc1Pathways)

hasDiffGenes<-!is.null(plotData$scesMerge@metadata$diffFC$genes)

hasDiffPathway<-!is.null(plotData$scesMerge@metadata$diffFC$pathways)

```

# Overview

**scRNABatchQC provides both metrics and diagnostic graphics to assess the similarity/difference on technical factors, expression profiles, and biological features across multiple scRNA-seq experiments, which helps identify potential batch effects. scRNABatchQC report includes five sections:**

  * _**QC Summary**_ - Overview of quality control in each scRNA-seq experiment. 
  * _**Technical View**_ - Technical features across multiple scRNA-seq experiments. 
  * _**Biological View**_ - Biological features across multiple scRNA-seq experiments.
  * _**Expression Similarity**_ - Expression similarity across multiple scRNA-seq experiments. 
  * _**Pairwise Difference**_ - Genes and pathways that drive the pairwise difference across multiple scRNA-seq experiments.

# QC Summary

```{r tableSummary, echo=FALSE, results="asis"}
kable(plotData$tableSummary, caption=tabRef("tableSummary", "Summary of Quality Metrics."), row.names=FALSE)
```

`r tabRef("tableSummary")` contains a summary of quality metrics generated by the scRNABatchQC package. 

A short description of Table 1 metrics is provided below: 

  * _**EID**_ - The name of the experiment
  * _**Count**_ - The total number of counts in the experiment 
  * _**Cell**_ - The total number of cells in the experiment
  * _**Gene**_ - The total number of genes in the experiment
  * _**R-Count**_ - The range of the total number of counts in a cell across all cells [min-median-max]. 
  * _**R-Gene**_ - The range of the total number of genes identified in a cell across all cells [min-median-max] 
  * _**mtRNA**_ - The maximum percentage of reads mapped to mitochondria-encoded genes in all cells 
  * _**rRNA**_ - The maximum percentage of reads mapped to genes encoded for ribosomal proteins in all cells
  * _**F-Count**_ - The number of cells filtered by the total number of counts 
  * _**C-Count**_ - The cutoff to remove the cell if the total counts of the cell is less than C-Count
  * _**F-Gene**_ - The number of cells filtered by the total number of detected genes
  * _**C-Gene**_ - The cutoff to remove the cell if the total genes of the cell is less than C-Gene
  * _**F-mt**_ - The number of cells filtered by the mtRNA percentage 
  * _**C-mt**_ - The cutoff to remove the cell if the percentage of mtRNA is larger than C-mt
  * _**F**_ - The total number of cells filtered by either counts, genes, or mtRNA criteria

# Technical View

**This section checks 11 technical features and aggregates results across multiple scRNA-seq experiments, which include the total number of counts, the number of expressed genes, percentage of reads mapped to mitochondria-encoded genes and ribosomal proteins, the expression cumulative plot, the detection rate, the mean-variance trend, and the variance explained by technical features.**

<hr>

```{r density_total_counts, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("density_total_counts", "The distribution of the total number of counts in a cell"), fig.height=4}
print(plotViolin(sces,feature="total_counts", featureLabel="The total number of counts", scolors=scolors, lineSize=plotData$lineSize))
```

`r figRef("density_total_counts")` shows the distribution of the total number of counts in a cell in each scRNA-seq experiment. Different experiments are denoted by different colors. Cells with very low number of read counts, suggesting RNAs are not efficiently captured, are considered as low quality and removed from the downstream analysis. Different count distributions across experiments are indicative of the presence of batch effects.

<hr>

```{r density_total_features, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("density_total_features", "The distribution of the total number of genes in a cell"), fig.height=4}
print(plotViolin(sces,feature="total_features", featureLabel="The number of expressed genes", scolors=scolors,lineSize=plotData$lineSize))
```

`r figRef("density_total_features")` shows the distribution of the number of expressed genes in a cell in each scRNA-seq experiment. Different experiments are denoted by different colors. Cells with very few expressed genes are considered as low quality and removed from the downstream analysis. Different distributions on gene numbers across experiments are indicative of the presence of batch effects.

<hr>

```{r density_pct_counts_Mt, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("density_pct_counts_Mt", "The distribution of percentage of mitochondrial counts in a cell"), fig.height=4}
print(plotViolin(sces,feature="pct_counts_Mt", featureLabel="The proportion of reads mapped to mitochondria-encoded genes", scolors=scolors,lineSize=plotData$lineSize))
```

`r figRef("density_pct_counts_Mt")` shows the distribution of the proportion of reads mapped to mitochondria-encoded genes in a cell in each scRNA-seq experiment. Different experiments are denoted by different colors. A cell with high proportion of reads mapped to mitochondria-encoded genes is considered as low quality and removed from the downstream analysis. Caution should be paid when one experiment has higher proportion of mitochondrial reads than others.

<hr>

```{r plotRibosomalRNA, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("plotRibosomalRNA", "The distribution of percentage of ribosomal counts in a cell"), fig.height=4}
print(plotViolin(sces,feature="pct_counts_rRNA", featureLabel="The proportion of reads mapped to genes encoded for ribosomal proteins", scolors=scolors,lineSize=plotData$lineSize))

```

`r figRef("plotRibosomalRNA")` shows the distribution of the proportion of reads mapped to genes encoded for ribosomal proteins in a cell in each scRNA-seq experiment, which is denoted by different color. Caution should be paid when one experiment has higher proportion of ribosomal gene reads than others.

<hr>

```{r genesCountdistribution, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("genesCountdistribution", "Cumulative expression plot for the top 500 highly expressed genes"), fig.height=4}

if(isInteractive){
  plotly::ggplotly(plotGeneCountDistribution(sces, scolors,lineSize=plotData$lineSize))
} else{
  print(plotGeneCountDistribution(sces, scolors,lineSize=plotData$lineSize)) 
}

```

`r figRef("genesCountdistribution")` shows the cumulative expression for the top 500 highly expressed genes. The plot is generated based on the average count of each gene, that is, the top 500 highly expressed genes are those with the highest average count. This plot describes how reads are distributed across genes in each scRNA-seq experiment. A cumulative graph with rapid increase at the beginning suggests a library with low complexity where very few genes consume most reads. In contrast, a cumulative graph with slow increase at the beginning indicates that reads are more evenly distributed across genes. Different cumulative expression plots across multiple scRNA-seq experiments are indicative of the presence of batch effects. 

<hr>

```{r aveCountVSdetectRate, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("aveCountVSdetectRate", "The detection rate of genes"), fig.height=4}

if(isInteractive){
  plotly::ggplotly(plotAveCountVSdetectRate(sces, scolors,lineSize=plotData$lineSize))  
} else {
  print(plotAveCountVSdetectRate(sces, scolors,lineSize=plotData$lineSize))
}

```

`r figRef("aveCountVSdetectRate")` shows the percentage of cells expressing each gene plotted against its log-average count. Generally the detection rate for a gene depends on its expression. Different detection rates across scRNA-seq experiments are indicative of the presence of batch effects. 

<hr>

```{r VarianceTrend, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("VarianceTrend", "The mean-variance trend")}
if(isInteractive){
  # plotly::ggplotly(plotVarianceTrend(sces, scolors,lineSize=plotData$lineSize))
} else {
  print(plotVarianceTrend(sces, scolors,lineSize=plotData$lineSize)) 
}
```

`r figRef("VarianceTrend")` shows mean vs. variance of gene expession. The variance of gene expression is driven by both uninteresting technical noise and real biological difference. scRNABatchQC models the technical variance of gene expression values by fitting a trend between mean expression abundance and technical variance. Different mean-variance trends across scRNA-seq experiments are indicative of the presence of batch effects.

<hr>

```{r log10_total_counts, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("log10_total_counts", "The percentage of expression variances explained by the library size"), fig.height=4}

if(isInteractive){
  plotly::ggplotly(plotVarianceExplained(sces, "genevar_by_counts", scolors,lineSize=plotData$lineSize))  
} else {
  print(plotVarianceExplained(sces, "genevar_by_counts", scolors,lineSize=plotData$lineSize))
}

```

`r figRef("log10_total_counts")` shows the percentage of gene expression variance explained by the log-transformed total number of counts. If percentages are generally small (1-3%), it indicates that gene expressions are not strongly associated with this technical factor. Otherwise, it suggests that the library sizes contribute substantially to the gene expression variance. If so, this technical factor may need to be modeled into the downstream analysis.  Different contributions of library sizes to expression variances across multiple experiments are indicative of the presence of batch effects.

<hr>

```{r log10_total_features, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("log10_total_features", "The percentage of expression variances explained by the total number of expressed genes"), fig.height=4}
if(isInteractive){
  plotly::ggplotly(plotVarianceExplained(sces, "genevar_by_features",scolors,lineSize=plotData$lineSize))  
} else {
  print(plotVarianceExplained(sces, "genevar_by_features",scolors,lineSize=plotData$lineSize)) 
}

```

`r figRef("log10_total_features")` shows the percentage of gene expression variance explained by the log-transformed number of genes detected in each cell. If the percentages are generally small (1-3%), it indicates that the gene expressions are not strongly associated with this technical factor. Otherwise, it suggests that the factor contribute substantially to the gene expression variance. If so, the technical factor may need to be modeled into the downstream analysis.  Different contributions of the factor to expression variances are indicative of the presence of batch effects.

<hr>

```{r log10_total_counts_Mt, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("log10_total_counts_Mt", "The percentage of expression variances explained by the proportion of mitochondrial reads"), fig.height=4}
if(isInteractive){
  plotly::ggplotly(plotVarianceExplained(sces,  "genevar_by_Mt", scolors,lineSize=plotData$lineSize))  
} else {
  print(plotVarianceExplained(sces,  "genevar_by_Mt", scolors,lineSize=plotData$lineSize))
}

```

`r figRef("log10_total_counts_Mt")` shows the percentage of expression variance of each gene that is explained by the percentage of mitochondrial reads. If the percentages are generally small (1-3%), it indicates that the gene expressions are not strongly associated with this technical factor. Otherwise, it suggests that the factor contribute substantially to the gene expression abundance. If so, the technical factor may need to be modeled into the downstream analysis.  Different contributions of the factor to expression variances are indicative of the presence of batch effects.

<hr>

```{r plotVariancerRNA, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("plotVariancerRNA", "The percentage of expression variances explained by the proportion of ribosomal proteins reads"), fig.height=4}
if(isInteractive){
  plotly::ggplotly(plotVarianceExplained(sces, "genevar_by_rRNA",scolors,lineSize=plotData$lineSize))  
} else {
  print(plotVarianceExplained(sces, "genevar_by_rRNA",scolors,lineSize=plotData$lineSize))
}

```

`r figRef("plotVariancerRNA")` shows the percentage of gene expression variance  explained by the percentage of reads mapped to ribosomal proteins. If the percentages are generally small (1-3%), it indicates that the gene expressions are not strongly associated with this technical factor. Otherwise, it suggests that the factor contribute substantially to the gene expression variance. If so, the technical factor may need to be modeled into the downstream analysis.  Different contributions of the factor to expression variances are indicative of the presence of batch effects.

<hr>

```{r plotSaturation, eval = hasSaturation, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap="Saturation curve", fig.height=4}
if(isInteractive){
  plotly::ggplotly(plotSaturation(plotData$saturation))
} else {
  print(plotSaturation(plotData$saturation))
}

```

<hr>


```{r child="scRNABatchQCreport_biological.Rmd", echo=FALSE}
```

```{r child="scRNABatchQCreport_similarity.Rmd", eval=hasComparison, echo=FALSE}
```

```{r plotPCA, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("plotPCA", "PCA plot of expression similarity between cells"), fig.height=4}
plotAllPCA(plotData$scesMerge, scolors, pointSize=plotData$pointSize)
```

`r figRef("plotPCA")` PCA Plot for the first two components, which shows the similarities between cells within or across scRNA-seq experiments.

<hr>

```{r plotTSNE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("plottSNE", "tSNE plot of expression similarity between cells"), fig.height=4}
plotAlltSNE(plotData$scesMerge, scolors, pointSize=plotData$pointSize)
```

`r figRef("plottSNE")` tSNE plot for the first two components, which shows the similarities between cells within or across scRNA-seq experiments. tSNE is a non-linear dimension reduction technique, which PCA reduce the dimension through linear transformation. 

<hr>


```{r child="scRNABatchQCreport_difference.Rmd", eval=hasComparison, echo=FALSE}
```

```{r child="scRNABatchQCreport_difference_pathways.Rmd", eval=hasComparison, echo=FALSE}
```

**Note**: If no pathways results are reported for HVGs, PC-related genes or differentially expressed genes, there are two possible reasons. One is that pathways are not enriched. The other reason is that gene name is not gene symbol or organism is not supported by WebGestalt (https://github.com/bzhanglab/WebGestaltR). WebGestalt supports 12 organisms, including athaliana, btaurus, celegans, cfamiliaris, drerio, sscrofa, dmelanogaster, ggallus, hsapiens, mmusculus, rnorvegicus, and scerevisiae.
